File: sentimental-credit/credit.py
from cs50 import get_string


def main():
    card_number = get_string("Number: ").strip()
    if check_sum(card_number):
        print(get_flag(card_number))

    print("INVALID")


# checking for the flag of the credit card
def get_flag(number):
    length = len(number)
    first_digit = int(number[0] + number[1])

    if length == 15:
        if first_digit == 34 or first_digit == 37:
            return "AMEX"

    elif length == 16:
        if first_digit in range(51, 56):
            return "MASTERCARD"
        if int(number[0]) == 4:
            return "VISA"

    elif length == 13 and int(number[0]) == 4:
        return "VISA"

    return "INVALID"


def check_sum(number):
    numbers = list(number)

    # multiply every other digit by 2
    every_other_digit = []
    reaming_digits = []

    numbers.reverse()

    for i in range(len(numbers)):
        if i % 2 == 0:
            reaming_digits.append(int(numbers[i]))
        else:
            product = int(numbers[i]) * 2
            if product > 9:
                every_other_digit.append(product - 9)
            else:
                every_other_digit.append(product)

    # add the sum to the sum of digits
    # that weren't multiplied by 2
    total_sum = sum(every_other_digit) + sum(reaming_digits)

    # if the total's last digit = 0
    # number is valid!
    if total_sum % 10 == 0:
        return True
    return False


main()

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>

string get_card_flag(string card_numbers);

int main(void)
{
    long card_number = get_long("Number: ");

    // isolate the numbers in characters to navegate between indexes
    char str[17];
    snprintf(str, sizeof(str), "%li", card_number);
    int len = strlen(str);

    // get the correct card flag
    string card_flag = get_card_flag(str);

    int count = 1;

    // product create to hold the multiplied values
    char product[17];
    product[0] = '\0';

    // loop trought the main string to get every other digit by 2 and multiply
    for (int i = len - 1; i >= 0; i--)
    {
        if ((len - i) % 2 == 0)
        {
            int digit = str[i] - '0';
            digit *= 2;

            char temp[3];
            snprintf(temp, sizeof(temp), "%d", digit);
            strcat(product, temp);

            memmove(&str[i], &str[i + 1], strlen(str) - i);
        }
        count++;
    }

    // loop for sum those products together
    int sum = 0;
    for (int i = 0; i < strlen(product); i++)
    {
        sum += product[i] - '0';
    }

    // loop for add to the sum of the digits that weren't multiplied by 2
    for (int i = 0; i < strlen(str); i++)
    {
        sum += str[i] - '0';
    }

    if (sum % 10 == 0)
    {
        printf("%s\n", card_flag);
    }
    else
    {
        printf("INVALID\n");
    }
}

// function responsible to deal with cards flags
string get_card_flag(string card_numbers)
{
    if (strlen(card_numbers) == 16)
    {
        if (card_numbers[0] == '5')
        {
            if (card_numbers[1] == '1' || card_numbers[1] == '2' || card_numbers[1] == '3' ||
                card_numbers[1] == '4' || card_numbers[1] == '5')
            {
                return "MASTERCARD";
            }
        }
        else if (card_numbers[0] == '4')
        {
            return "VISA";
        }
    }
    else if (strlen(card_numbers) == 15)
    {
        if (card_numbers[0] == '3')
        {
            if (card_numbers[1] == '4' || card_numbers[1] == '7')
            {
                return "AMEX";
            }
        }
    }
    else if (strlen(card_numbers) == 13)
    {
        if (card_numbers[0] == '4')
        {
            return "VISA";
        }
    }
    return "INVALID";
}
